src/event_listener.rs

```diff
@@ -0,0 +1,194 @@
+    /// no-op.
+    ///
+    /// Note that the this function must be implemented in a way such that
+    /// it should not run for an extended period of time before the function
+    /// returns.  Otherwise, RocksDB may be blocked.
+    fn on_flush_completed(&self, _flush_job_info: &FlushJobInfo) {}
+}
+
+/// Flush information
+#[derive(Debug)]
+pub struct FlushJobInfo {
+    /// The name of the column family
+    pub cf_name: String,
+    /// The path to the newly created file
+    pub file_path: String,
+    /// The smallest sequence number in the newly created file
+    pub smallest_seqno: u64,
+    /// The largest sequence number in the newly created file
+    pub largest_seqno: u64,
+    /// Flush reason
+    pub flush_reason: FlushReason,
+}
+
+/// Flush reason
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[repr(i32)]
+pub enum FlushReason {
+    Others = 0,
+    GetLiveFiles = 1,
+    ShutDown = 2,
+    ExternalFileIngestion = 3,
+    ManualCompaction = 4,
+    WriteBufferManager = 5,
+    WriteBufferFull = 6,
+    Test = 7,
+    DeleteFiles = 8,
+    AutoCompaction = 9,
+    ManualFlush = 10,
+    ErrorRecovery = 11,
+    ErrorRecoveryRetryFlush = 12,
+    WalFull = 13,
+    CatchUpAfterErrorRecovery = 14,
+}
+
+struct EventListenerCallback<T>
+where
+    T: EventListener + 'static + Send + Sync,
+{
+    inner: Arc<T>,
+}
+
+impl<T> EventListenerCallback<T>
+where
+    T: EventListener + 'static + Send + Sync,
+{
+    pub fn new(listener: Arc<T>) -> Self {
+        Self { inner: listener }
+    }
+
+    pub unsafe extern "C" fn destructor(raw_cb: *mut c_void) {
+        drop(Box::from_raw(raw_cb as *mut Self));
+    }
+
+    pub unsafe extern "C" fn name(raw_cb: *mut c_void) -> *const libc::c_char {
+        let cb = &*(raw_cb as *mut Self);
+        cb.inner.name().as_ptr()
+    }
+
+    pub unsafe extern "C" fn on_flush_completed(
+        raw_cb: *mut c_void,
+        _db: *mut rocksdb_t,
+        flush_job_info: *const rocksdb_flushjobinfo_t,
+    ) {
+        let cf_name_ptr = ffi::rocksdb_flushjobinfo_cf_name(flush_job_info);
+        let cf_name = if cf_name_ptr.is_null() {
+            String::new()
+        } else {
+            let name = from_cstr(cf_name_ptr);
+            ffi::rocksdb_free(cf_name_ptr as *mut c_void);
+            name
+        };
+
+        let file_path_ptr = ffi::rocksdb_flushjobinfo_file_path(flush_job_info);
+        let file_path = if file_path_ptr.is_null() {
+            String::new()
+        } else {
+            let path = from_cstr(file_path_ptr);
+            ffi::rocksdb_free(file_path_ptr as *mut c_void);
+            path
+        };
+
+        let smallest_seqno = ffi::rocksdb_flushjobinfo_smallest_seqno(flush_job_info);
+        let largest_seqno = ffi::rocksdb_flushjobinfo_largest_seqno(flush_job_info);
+        let flush_reason = std::mem::transmute::<libc::c_int, FlushReason>(
+            ffi::rocksdb_flushjobinfo_flushreason(flush_job_info),
+        );
```

Comment:
I based this on how it's done in `Options::logger_callback`; the alternative is to match the value and translate but I hit some snags with `bindgen`-generated constants having different signedness on Linux/macOS vs Windows.

Reply:
In that case, there is really no need for the enum defined in the c header file, right?